LeaderboardBackend/Services/Impl/UserService.cs

```diff
@@ -35,9 +34,41 @@ public UserService(ApplicationContext applicationContext)
         );
     }
 
-    public async Task CreateUser(User user)
+    public async Task<CreateUserResult> CreateUser(RegisterRequest request)
     {
-        _applicationContext.Users.Add(user);
+        /* Linq query translated as:
+        SELECT count(*) FILTER (WHERE lower(t.username) = @__ToLower_0)::int > 0 AS "usernameExists",
+            count(*) FILTER (WHERE lower(t.email) = @__ToLower_1)::int > 0 AS "emailExists"
+        FROM (
+            SELECT u.email, u.username, 1 AS "Key"
+            FROM users AS u
+        ) AS t
+        GROUP BY t."Key"
+        LIMIT 1
+        */
+        var conflicts = await _applicationContext.Users.GroupBy(x => 1)
```

Comment:
Without the `GroupBy`, the parameter for  the `Select` lambda would be a single `User` entity, so it would not be possible to call `Count()` on it.
It can be done without the `GroupBy` this way: 
```csharp
var conflicts = await _applicationContext.Users
.Select(users => new
{
    usernameExists = _applicationContext.Users.Count(x => x.Username.ToLower() == request.Username.ToLower()) > 0,
    emailExists = _applicationContext.Users.Count(x => x.Email.ToLower() == request.Email.ToLower()) > 0
}).FirstOrDefaultAsync();
```
_(note the repeated use of `_applicationContext.Users`)_
which produces the following SQL query:
```sql
SELECT (
    SELECT count(*)::int
    FROM users AS u0
    WHERE lower(u0.username) = @__ToLower_0) > 0 AS "usernameExists", (
    SELECT count(*)::int
    FROM users AS u1
    WHERE lower(u1.email) = @__ToLower_1) > 0 AS "emailExists"
FROM users AS u
LIMIT 1
```

The current SQL query for reference:
```sql
SELECT count(*) FILTER (WHERE lower(t.username) = @__ToLower_0)::int > 0 AS "usernameExists",
    count(*) FILTER (WHERE lower(t.email) = @__ToLower_1)::int > 0 AS "emailExists"
FROM (
    SELECT u.email, u.username, 1 AS "Key"
    FROM users AS u
) AS t
GROUP BY t."Key"
LIMIT 1
```

I don't know which SQL query is better (if there is even a difference), I just think the code with the `GroupBy` is more readable because it doesn't repeatedly use `_applicationContext.Users`.

Reply:
I see. Thanks for the detailed info

Let's keep the `GroupBy` :+1: